### ISSUE TITLE
Write and run strong scaling jobs

### ISSUE THREAD

**Issue by lyon-fnal** [lyon-fnal/IRMA#3]
Write and run strong scaling jobs that make one histogram of cluster energy.

**Comment by lyon-fnal** [comment]
Runs on my Mac - ready to try on Cori

**Comment by lyon-fnal** [comment]
See `jobs/003_StrongScaling`

**Comment by lyon-fnal** [comment]
I ran a 1x32 (1 node, 32 tasks/node) job and it died with out of memory. I don't think that should have happened. Need to check how much memory usage is going on.

**Comment by lyon-fnal** [comment]
I'm manually reading 526,125,917 rows (that's what the 1x32 job tried to do) with `@time`. Since the data are single precision floats, that should be 2.1 GB. And it is!

```julia
julia> @time energyData = energyDS[1, s:e]
 62.975539 seconds (1.37 M allocations: 2.028 GiB, 0.05% gc time)
```

Doing more memory checking with `Base.gc_bytes()`. It seems that at the start of the script, there are 514 MB already allocated. This is after loading all packages and initializing MPI. After compiling `histogramEnergy`, it jumps to 650 MB. Just getting into the `go()` function brings it up to 875 MB, except for rank 0, which jumps to 1.4 GB! I think this is because `go()` is compiled by rank 0 and it lives in the cache for the other ranks. By the time it gets to the dataset read, most ranks are at 895 MB, but rank 0 is at 1.45 GB. Two of the ranks made it past the read before the memory ran out. They hit 3.015 GB. That's about right for a 2 GB read. 

So we can see we're close to the edge of running out of memory.

**Comment by lyon-fnal** [comment]
OK - so it looks like the large memory use is coming from the JIT compilation of the `go()` function (rank 0 actually does the compilation - using lots of memory - and the other ranks get it from some cache). I can run with the `--compile=min` option (see https://github.com/JuliaLang/julia/issues/36017), which only compiles modules, and the memory use is much less. But the code runs way more slowly (especially filling the histogram, which appears to be glacial). 

Indeed with `--compile=min`, the job starts off at 353MB and grows only to 358MB before it loads the data, after which it grows to 2.5 GB. But filling the histogram effectively takes forever. 

I think the right thing to do is to create a module for much of the code and be sure that gets precompiled. Read https://docs.julialang.org/en/v1/manual/modules/#Module-initialization-and-precompilation for how to do that. 

I can probably make this even more efficient by using the package compiler too (see https://github.com/JuliaLang/PackageCompiler.jl).

**Comment by lyon-fnal** [comment]
So it looks like the way I need to do this is to put the code for the job inside the IRMA package source structure as a submodule of IRMA. Then it will get precompiled with IRMA itself.

**Comment by lyon-fnal** [comment]
Well - very strange. I tried putting much of the code into the IRMA package, but the memory usage got worse! Loading IRMA made memory jump from 84MB to 809MB. Then just entering the function `go` made the memory jump up to 1.5GB !! 

Removing most of go and adding things back bit by bit, memory in go jumped when I added back MPI stuff.

So let's try this ... remove go from IRMA (it didn't help). Make a script where most of the MPI stuff is in the global space. Let's see what that does. In the meantime, I really need a way to get memory allocation easily. Perhaps with `Cassette.jl`

**Comment by lyon-fnal** [comment]
So I think I've convinced myself that the absolute value of `gc_bytes` is not useful for determining the instantaneous memory usage of the Julia process. Using instruments on the Mac, I find that for a two process MPI run, each Julia process takes at most about 500 MB (usually was around 300 MB). Using `time -l` on the `mpiexec`, it says that the maximum resident memory usage was about 700 MB. I assume that's for the whole thing. 

I'm trying to see if Instruments will plot memory usage for just one process, but it doesn't seem to do that. 

Anyway, maybe differences reported by `gc_bytes` are useful, but not the absolute value. That being said, 32 processes at about 500 MB each adds up to 16 GB! That's without loading data.

**Comment by lyon-fnal** [comment]
**RESULTS**

I've run `jobs/003_StrongScaling/run_strongScalingJob.sh` (which runs `strongScalingJob.jl` in that directory) from commit 82b715b to make,
```
lyon@cori03:/global/cscratch1/sd/lyon/003_StrongScaling> ls
histos_10x32.jld2  histos_2x32.jld2  histos_3x32.jld2  histos_4x32.jld2  histos_5x32.jld2  histos_6x32.jld2  histos_7x32.jld2  histos_8x32.jld2  histos_9x32.jld2
```

The log files are in `jobs/003_StrongScaling` for now.

**Comment by lyon-fnal** [comment]
There's still a problem when running on one node and running out of memory. Look at that in a new issue.

**Comment by lyon-fnal** [comment]
See #19 for the out of memory problems.

**Comment by lyon-fnal** [comment]
Re-opening this issue. Explore Julia startup and timing. Maybe use PackageCompiler.jl .

**Comment by lyon-fnal** [comment]
Use `PackageCompiler.jl` .  

- [x] Note that I need to re-run 5x32 since I ran that with part-way compiled code.  -- No - I'm not going to run it again. I have it on my Mac. 

See `Clima/.dev` for a compilation script.

**Comment by lyon-fnal** [comment]
Opened PR #22 with changes for running with the Package Compiler. See `compile.jl`. Compilation took about 4 minutes. 

I ran the jobs in `run.sh`. (the `-W` flag to `sbatch` tells it to wait until the job is done before returning. That worked nicely for submitting a series of jobs unattended. Note that the 5x1 job failed because it exceeded the 5 minute time limit I had.

I ran jobs with commit ff854981 on `lyon-fnal/issue3`
